### ISSUE TITLE
apiserver: command exec in pod randomly drops stdin

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#112834]
### What happened?

I was using https://github.com/kubernetes/kubernetes/blob/725993afce11082b973901d2e92553640771d55d/test/e2e/framework/exec_util.go#L49-L84 (similar to `kubectl exec -i`) with `dd` as command and some data as stdin to create a file inside a container. This worked fine most of the time during local development. When used inside a Prow job, I ran into test flakes that could be traced down to `dd` returning after writing zero bytes although stdin was non-empty.

### What did you expect to happen?

The data should be transferred reliably.

### How can we reproduce it (as minimally and precisely as possible)?

Sorry, no test for this right now.

### Kubernetes version

462f41212a4e5ded8974784038f681a08dded1be (roughly current master)

### Cloud provider

<details>
kind
</details>

### Install tools

kind

### Container runtime (CRI) and version (if applicable)

containerd (from master)

**Comment by aojea** [comment]
maybe something breaking the spdy encapsulation? 
it can be between e2e -- apiserver --- kubelet -- containerd

**Comment by pohly** [comment]
Yes, could be anywhere. Probably a good first step would be to write a stress test for this functionality.

**Comment by aojea** [comment]
yeah, it suffered of memory leaks in the past, I don't think that this feature has been stress tested

**Comment by pohly** [comment]
/sig apimachinery

This is used by test/e2e/framework, but ultimately this is caused somewhere else.

/remove-lifecycle stale

**Comment by cici37** [comment]
We are at api machinery bugscrub meeting. 
Thanks for the tagging! Looks like the root cause hasn't been identified and [test/e2e/storage/drivers/proxy/io.go](https://github.com/DrAuYueng/kubernetes/commit/309e577d26957e6575dc1993b33c27ea27f65476#diff-b85441025e78b8f9e9a08d916a51a5e6007836c4fd5678bb77b8fd5e68096ba7) could be used to reproduce?
@jpbetz Would you mind to take a look or help with redirecting to the right person? Thanks

**Comment by pohly** [comment]
I have asked a colleague to write a stress test for this. He won't be able to solve it, but it should be a start towards identifying the root cause.

/cc @hj-johannes-lee

**Comment by aojea** [comment]
the tests in `pkg/kubelet/cri/streaming/server_test.go` may serve as base

**Comment by cici37** [comment]
/cc @hj-johannes-lee @pohly 
Would you mind sharing if the stress test is planned or if there is any other plan on this? Thank you!

**Comment by hj-johannes-lee** [comment]
@cici37 Hi, I am working on it, but could not reproduce the exact problem.! We were supposed to have a meeting and discuss this matter, but we couldn't.
What I did until now is as follows:
1. Deploying a busybox pod.
2. Using ExecWithOptions with stdin option (here, data is read from a file that has a long enough contents) for creating a file inside a container.
3. Comparing the value of stdout read from the container to the value of data read from the file.
4. Looping the previous process so it will continue until the case is caught.

Let me discuss with @pohly asap and come with the "working" stress test.!

**Comment by hj-johannes-lee** [comment]
@cici37  @pohly 
I just submitted a pr, but I know it should be improved in many ways.!

**Comment by shawkins** [comment]
We in the fabric8 java client are seeing similar problems with exec stdin across a variety of our supported http clients.  We're also seeing a problem the other direction on stdOut.  Occasionally we'll see the web socket close being sent to before all of the stdOut data messages have been received.  We are considering a workaround of waiting for the exit code message, but it is definitely not expected behavior to see additional messages after close.

**Comment by aojea** [comment]
websocket has a known issue https://github.com/kubernetes/kubernetes/issues/89899#issuecomment-1143026673 and is reliable than the spdy implementation

**Comment by shawkins** [comment]
@aojea that known issue is fairly easy to workaround, at least for us.  When you are done with stdIn, you send the web socket close.  Isn't SPDY deprecated?

The problem identified here is that the apiserver websocket seems to be the handling close out of order - when sending data it will process the incoming close prior to fully processing data messages it has already received, and on the send side it will send the close ahead of pending data messages.

**Comment by aojea** [comment]
> Isn't SPDY deprecated?

https://github.com/kubernetes/kubernetes/issues/7452

but is the one used by kubectl exec/portforward 😄 

There was one person trying to move on to websockets by default but it seems that is hard to keep the compatibility between versions https://github.com/kubernetes/kubernetes/issues/60140#issuecomment-1411782810 and the ROI doesn't seem to be enough to try such complicated change


> The problem identified here is that the apiserver websocket seems to be the handling close out of order - when sending data it will process the incoming close prior to fully processing data messages it has already received, and on the send side it will send the close ahead of pending data messages.

the close before the data is sent is similar to this https://github.com/kubernetes/kubernetes/issues/60140#issuecomment-1364264111, but I'm just doing some quick correlations here , you should verify and double check those links

**Comment by shawkins** [comment]
> the close before the data is sent is similar to this https://github.com/kubernetes/kubernetes/issues/60140#issuecomment-1364264111, but I'm just doing some quick correlations here , you should verify and double check those links

My apologies.  After reviewing these issues and the relevant net, apiserver, and kubernetes code, the issues that we are hitting are definitely within the fabric8 kubernetes client code.  I'll get those addressed and see how often, if at all, we're actually hitting this or #60140

**Comment by XciD** [comment]
Hello,

Not sure it's the same issue. I don't succeed to use `dd` with stdin with kube exec.
```
echo 'test' > file
cat file | kubectl exec -i pod -- dd of=/home/user/file
kubectl exec pod -- ls -lh /home/user/file
-rw-r--r-- 1 user user 0 Apr 11 22:19 /home/user/file
```

**Comment by aojea** [comment]
This needs to be updated with more recent information and more specific, lots of things changed since this was reported

**Comment by pohly** [comment]
/close

@XciD: please open a new issue.
